### PR TITLE
Add Repl docs

### PR DESCRIPTION
**Pull request by 0xKanekiKen**
## Describe your changes
This PR updates the docs with the repl cli and commands. 

## Checklist before requesting a review
- Repo forked and branch created from `next` according to naming convention.
- Commit messages and codestyle follow [conventions](./CONTRIBUTING.md).
- Relevant issues are linked in the PR description.
- Tests added for new functionality.
- Documentation/comments updated according to changes.

### CODE REVIEW

File: docs/src/intro/usage.md [hunk 1]
```diff
@@ -36,6 +36,7 @@ Currently, Miden VM can be executed with the following subcommands:
 * `verify` - this will verify a previously generated proof of execution for a given program.
 * `compile` - this will compile a Miden assembly program (i.e., build a program [MAST](../design/programs.md)) and outputs stats about the compilation process.
 * `analyze` - this will run a Miden assembly program against specific inputs and will output stats about its execution.
+* `repl` - this will initiate the Repl tool for Miden VM
```
Comment (bobbinth) [inline docs/src/intro/usage.md:39]: Let's change this to: "this will initiate the Miden REPL tool." and make "REPL tool" a link to the new section below.

File: docs/src/intro/usage.md [hunk 2]
```diff
@@ -51,4 +52,138 @@ In the `miden/examples/fib` directory, we provide a very simple Fibonacci calcul
 ```
 ./target/release/miden run -a miden/examples/fib/fib.masm -n 1
 ```
-This will run the example code to completion and will output the top element remaining on the stack.
\ No newline at end of file
+This will run the example code to completion and will output the top element remaining on the stack.
+
+## REPL
+
+The Miden Read–eval–print loop or repl for short is a Miden shell that allows for quick and easy debugging with Miden assembly. When the repl gets initialised, there are a bunch of awesome things an user can do like execute any Miden instruction, undo executed instructions, check the state of the stack and memory at anytime and more! A `history.txt` file is saved when the repl is exited. One thing to note that all the repl native commands starts with an `!` to differentiate them with regular assembly instruction. The Repl currently supports the following commands:
```
Comment (bobbinth) [inline docs/src/intro/usage.md:59]: nit: let's capitalize "REPL" here and in other places - i.e., "repl" -> "REPL".

Also, there are a few types and grammatical mistakes in this paragraph. I rewrote it - so, let's use the version below:

The Miden Read–eval–print loop (REPL) is a Miden shell that allows for quick and easy debugging of Miden assembly. After the REPL gets initialized, you can use it execute any Miden instruction, undo executed instructions, check the state of the stack and memory at a given point in time and do many other useful things! A `history.txt` file is saved when the REPL is exited. One thing to note that all the REPL native commands starts with an `!` to differentiate them from regular assembly instructions. The REPL currently supports the following commands:

File: docs/src/intro/usage.md [hunk 3]
```diff
@@ -51,4 +52,138 @@ In the `miden/examples/fib` directory, we provide a very simple Fibonacci calcul
...
+
+### Miden assembly instruction
+
+All the Miden instruction mentioned in the Miden Assembly document is valid. (Ex. push.0, drop, dropw, swap, cswap, u32checked_add, mem_loadw, mtree_get etc.). One can either input instructions one by one or multiple instructions in one input.
```
Comment (bobbinth) [inline docs/src/intro/usage.md:63]: A couple typos here:

* "All Miden instructions ..."
* "... document are valid." Also, the "document" should probably be "section" and we should probably use a link for it.

File: docs/src/intro/usage.md [hunk 4]
```diff
@@ -51,4 +52,138 @@ In the `miden/examples/fib` directory, we provide a very simple Fibonacci calcul
...
+>> push.1
+>> push.2
+>> push.3
+```
+
+```
+push.1 push.2 push.3
+```
+
+In order to execute a control flow operation, one needs to write the entire flow operation in a single line with spaces between individual operations.
```
Comment (bobbinth) [inline docs/src/intro/usage.md:76]: "entire flow operation" -> "entire statement".

File: docs/src/intro/usage.md [hunk 5]
```diff
@@ -51,4 +52,138 @@ In the `miden/examples/fib` directory, we provide a very simple Fibonacci calcul
...
+If the `addr` has not been initialied:
+
+```
+>> !mem[87]
+Memory at address 87 is empty
+```
+
+### !undo
+
+`!undo` command reverts to the previous state of the stack and memory by dropping off the last executed assembly instruction from the program. One could use `undo` as many times as they want to restore the state of a stack and memory n instructions ago(provided there are n instructions in the program). If there are no instructions in the repl tool and `!undo` command is sent as an input then it will result in an error.
```
Comment (bobbinth) [inline docs/src/intro/usage.md:172]: * Missing space after "ago".
* Let's write `n` as `$n$` so that it appears in math notation.
* In one case `undo` does not start with `!`.

File: docs/src/intro/usage.md [hunk 6]
```diff
@@ -51,4 +52,138 @@ In the `miden/examples/fib` directory, we provide a very simple Fibonacci calcul
...
+If the memory is not yet been initialised, then the commands prints that the memory is not yet initialsed.
+
+```
+> !mem
+The memory has not been initialised yet
+```
+
+### !mem[addr]
+
+`!mem[addr]` prints the memory value(in integer) present in the address `addr`.
```
Comment (bobbinth) [inline docs/src/intro/usage.md:154]: Let's change this to:

`!mem[addr]` command prints out memory contents at the address specified by `addr`.

File: miden/src/repl/mod.rs [hunk 7]
```diff
@@ -245,8 +250,8 @@ fn execute(program: String) -> Result<(Vec<(u64, Word)>, Vec<Felt>), MidenError>
 
     let pub_inputs = vec![];
     let inputs = ProgramInputs::new(&pub_inputs, &[], vec![]).unwrap();
-    let mut process = Process::new_debug(inputs);
-    let _program_outputs = process
+    let mut process = Process::new_debug(program.kernel(), inputs);
+    let _program = process
```
Comment (bobbinth) [inline miden/src/repl/mod.rs:254]: Why did you change this from `_program_outputs` to `_program`? I think the previous name was correct.

File: miden/src/repl/mod.rs [hunk 8]
```diff
@@ -299,14 +304,14 @@ fn print_instructions() {
     println!("!mem[i]: displays the state of the memory at address i");
     println!("!undo: remove the last instruction");
     println!("!program: display the program");
-    println!("!help: prints out all the available commands");
+    println!("!help: prints out all the available commands\n");
```
Comment (bobbinth) [inline miden/src/repl/mod.rs:307]: I think this is the only place where we should add a newline (all other added newlines are not needed). But I would add it as a separate `println!();` to keep it consistent with how we did it a few lines before.

File: docs/src/intro/usage.md [hunk 9]
```diff
@@ -51,4 +52,138 @@ In the `miden/examples/fib` directory, we provide a very simple Fibonacci calcul
...
+begin
+    push.1.2.3.4
+    repeat.16 pow2 end
+    u32checked_add
+end
+```
+
+### !stack
+
+`!stack` command prints out the state of the stack at the latest clock cycle. The state also contains the overflow part of the stack. The stack will be empty at initialisation i.e. it will only have 16 `0`.
```
Comment (bobbinth) [inline docs/src/intro/usage.md:113]: Change to:

`!stack` command prints out the state of the stack at the last executed instruction. Since the stack always contains at least 16 elements, 16 or more elements will be printed out (even if all of them are zeros).

File: docs/src/intro/usage.md [hunk 10]
```diff
@@ -51,4 +52,138 @@ In the `miden/examples/fib` directory, we provide a very simple Fibonacci calcul
...
+
+The `!stack` command will print out the following state of the stack:
+```
+>> !stack
+3072 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
+```
+
+### !mem
+
+`!mem` command prints the entire state of the memory which has been initialised. For all the values memory addresses which has been initialised, it prints the address along with it's memory register's value (integer).
```
Comment (bobbinth) [inline docs/src/intro/usage.md:132]: Change to:

`!mem` command prints out contents of all initialized memory locations. For each such location, the address along with its memory values are printed. Recall that 4 elements are stored at each memory address.

File: docs/src/intro/usage.md [hunk 11]
```diff
@@ -51,4 +52,138 @@ In the `miden/examples/fib` directory, we provide a very simple Fibonacci calcul
...
+If the memory has at least one value which has been initialised:
+
+```
+>> !mem
+7: [1, 2, 0, 3]
+8: [5, 7, 3, 32]
+9: [9, 10, 2, 0]
+```
+
+Here 7, 8 and 9 addresses have been initialised and their respective counter-parts in the same line are the memory values on these addresses.
```
Comment (bobbinth) [inline docs/src/intro/usage.md:143]: Let's delete this line.

File: docs/src/intro/usage.md [hunk 12]
```diff
@@ -51,4 +52,138 @@ In the `miden/examples/fib` directory, we provide a very simple Fibonacci calcul
...
+
+If the memory is not yet been initialised, then the commands prints that the memory is not yet initialsed.
```
Comment (bobbinth) [inline docs/src/intro/usage.md:145]: Let's change "initialised" to "initialized" here and everywhere else.

**Review by bobbinth — CHANGES_REQUESTED** [review]
Thank you! I left some comments inline.

File: docs/src/intro/usage.md (hunk 2 above)
Comment (0xKanekiKen) [reply docs/src/intro/usage.md]: Thanks! I've capitalized all the ""repl" in this section and reworked on the REPL introduction. Let me know your thoughts on the current definition.

File: docs/src/intro/usage.md (hunk 5 above)
Comment (0xKanekiKen) [reply docs/src/intro/usage.md]: Fixed, thanks!

File: miden/src/repl/mod.rs (hunk 7 above)
Comment (0xKanekiKen) [reply miden/src/repl/mod.rs]: Reverted it back to old name.

File: docs/src/intro/usage.md [hunk 13]
```diff
@@ -51,4 +52,137 @@ In the `miden/examples/fib` directory, we provide a very simple Fibonacci calcul
...
+repeat.20 pow2 end
+```
+
+### !help
+
+The `!help` command prints out all the available commands in the REPL tool. 
+
+### !program
+
+The `!program` command prints out the entire miden program getting executed. E.g., in the below scenario:
```
Comment (bobbinth) [inline docs/src/intro/usage.md:96]: nit: "Miden" should be capitalized.

I would also say "being executed" rather than "getting executed."

File: docs/src/intro/usage.md [hunk 14]
```diff
@@ -51,4 +52,137 @@ In the `miden/examples/fib` directory, we provide a very simple Fibonacci calcul
...
+```
+
+### !mem[addr]
+
+The `!mem[addr]` command prints out memory contents at the address specified by `addr`.
+
+If the `addr` has been initialized:
+
+```
+>> `!mem[9]`
```
Comment (bobbinth) [inline docs/src/intro/usage.md:158]: nit: we probably don't need the quotation marks here.

**Review by bobbinth — CHANGES_REQUESTED** [review]
Thank you! Looks good! Just a couple of small comments inline.

**Review by bobbinth — APPROVED** [review]
All looks good! Thank you!
